core/src/main/java/org/apache/carbondata/core/scan/executor/impl/AbstractQueryExecutor.java

```diff
@@ -508,7 +508,8 @@ private BlockExecutionInfo getBlockExecutionInfoForBlock(QueryModel queryModel,
     int[] dimensionChunkIndexes = QueryUtil.getDimensionChunkIndexes(projectDimensions,
         segmentProperties.getDimensionOrdinalToChunkMapping(),
         currentBlockFilterDimensions, allProjectionListDimensionIndexes);
-    ReusableDataBuffer[] dimensionBuffer = new ReusableDataBuffer[projectDimensions.size()];
+    // 4 because each column(including itself) can have a maximum of 3 nested complex levels
+    ReusableDataBuffer[] dimensionBuffer = new ReusableDataBuffer[projectDimensions.size() * 4];
```

Comment:
Should calculate the total dimensions when flattened and use it. 
always hardcoding to 4 is bad as if I have 100 primitive column, I waste 300 more.

calculate in init method and keep a class variable.

Reply:
done

Reply:
done